sihl/test/database_migration.ml

```diff
@@ -1,90 +1,113 @@
+;;
+
+module Make (MigrationService : Sihl.Contract.Migration.Sig) = struct
+  let run_migrations _ () =
+    let open Lwt.Syntax in
+    let* () = clean_state () in
+    (* Use test migration state table to not interfere with other tests *)
+    Sihl.Configuration.store
+      [ "MIGRATION_STATE_TABLE", "test_core_migration_state" ];
+    let* () = MigrationService.lifecycle.stop () in
+    let* () = MigrationService.lifecycle.start () in
+    let migration : Sihl.Database.Migration.t =
+      ( "professions"
+      , [ Sihl.Database.Migration.create_step
+            ~label:"create table"
+            "CREATE TABLE professions (salary INT);"
+        ] )
+    in
+    let* () = MigrationService.execute [ migration ] in
+    let* _ =
+      (* If this doesn't fail, test is ok *)
+      Sihl.Database.query (fun (module Connection : Caqti_lwt.CONNECTION) ->
+          Connection.find_opt
+            (Caqti_request.find_opt
+               Caqti_type.unit
+               Caqti_type.int
+               "SELECT salary FROM professions;")
+            ())
+      |> Lwt.map Sihl.Database.raise_error
+    in
+    Lwt.return ()
+  ;;
+
+  let pending_migrations _ () =
+    let open Lwt.Syntax in
+    let* () = clean_state () in
+    (* Use test migration state table to not interfere with other tests *)
+    Sihl.Configuration.store
+      [ "MIGRATION_STATE_TABLE", "test_core_migration_state" ];
+    let* () = MigrationService.lifecycle.stop () in
+    let* () = MigrationService.lifecycle.start () in
+    let* status = MigrationService.pending_migrations () in
+    Alcotest.(check (list (pair string int)) "no pending migrations" [] status);
+    let migration1 : Sihl.Database.Migration.t =
+      ( "professions"
+      , [ Sihl.Database.Migration.create_step
+            ~label:"create professions table"
+            "CREATE TABLE professions (salary INT);"
+        ] )
+    in
+    MigrationService.register_migrations [ migration1 ];
+    let* status = MigrationService.pending_migrations () in
+    Alcotest.(
+      check
+        (list (pair string int))
+        "one pending migration"
+        [ "professions", 1 ]
+        status);
+    let* () = MigrationService.run_all () in
+    let migration2 : Sihl.Database.Migration.t =
+      ( "orders"
+      , [ Sihl.Database.Migration.create_step
+            ~label:"create orders table"
+            "CREATE TABLE orders (amount INT);"
+        ] )
+    in
+    MigrationService.register_migrations [ migration2 ];
+    let* status = MigrationService.pending_migrations () in
+    Alcotest.(
+      check
+        (list (pair string int))
+        "one pending migration"
+        [ "orders", 1 ]
+        status);
+    let* () = MigrationService.run_all () in
+    let* status = MigrationService.pending_migrations () in
+    Alcotest.(check (list (pair string int)) "desired state reached" [] status);
+    Lwt.return ()
+  ;;
+
+  let suite =
+    Alcotest_lwt.
+      [ ( "migration"
```

Comment:
Please also add a test where the migration state is ahead of the registered migrations. This should result in an empty list for pending migrations. On this note: Who is now responsible for reporting that a user should not delete migrations as they are append only?

Reply:
This is really hard to test with the current API, we need a way to clean the registered migrations just to write this test. It is also really hard to assert on logs.I left your warnings so the user gets notified just like they did before. 

Maybe we should look into a way for writing tests for private API in the future.

Reply:
You could just increment the `core_migration_status` table entry directly with SQL without having any registrations registered. What do you think?

Also, as mentioned before, since you now do this if-check
```ocaml
      | (namespace, Some n) :: xs ->
        if n > 0
```
you automatically ignore all cases where deletion of migrations could have occured (unless I misunderstood something)

Reply:
That's true, I could add the test but it is not possible to assert on anything here.

This is the API under test:

```ocaml
  (** [pending_migrations ()] returns a list of migrations that need to be
      executed in order to have all migrations applied. The returned migration
      is a tuple [(namespace, number)] where [namespace] is the namespace of the
      migration and [number] is the number of pending migrations that need to be
      applied in order to achieve the desired schema version.

      An empty list means that there are no pending migrations and that the
      database schema is up-to-date. *)
  val pending_migrations : unit -> (string * int) list Lwt.t
```

We don't expose the information about deleted migrations to the end user.

Reply:
The only questions this API wants to answer are:

- are there any migrations pending?
- if so, which ones?

Reply:
Ah I now see that all warnings are still displayed as before. I overread 
> I left your warnings so the user gets notified just like they did before.

Thanks!

Reply:
Then it's all good